### DISCUSSION TITLE
presence_penalty

### DISCUSSION THREAD

**Discussion by SwHaraday** [QwenLM/Qwen3#1744] [Q&A] [answered]
Hi,
I'm playing with Qwen3-4B-Instruct.
It is really impressive compare to others. However it some times gives answer with same sentence over and over again.

They say 'presence_penalty=1.5' will make it better.
Now the question is what is the range for this parameter ?
for example from 0.0 to 100.0

And what is going to happen if I set 'presence_penalty=50.0' ?

Thank you in advance.

**Comment by jklj077** [comment, marked as answer]
The neutral value for `presence_penalty` is 0.0 and the most API providers limit its range to [-2.0, 2.0]. 

> And what is going to happen if I set 'presence_penalty=50.0' ?

In most implementations, the `presence_penalty` is subtracted from the logits (the raw, unnormalized scores) of any token that has already appeared in the input context. A very high value like 50.0 effectively drives the logits of those tokens far into the negative, making their probabilities negligible after the softmax transformation.

As a result, the model will almost certainly avoid repeating any token that has already occurred in the prompt or generated output. While this can strongly discourage repetition, it may also lead to unnatural or incoherent text, especially if the model is forced to avoid common or contextually necessary words.

**Reply by SwHaraday** [reply]
Hi jklj077,
thank you for your detailed explanation.
i've tried several settings and boiled down to 1.9 . now works nicely.

btw is there docs related to this kind of parameters and explanations ? if exists, guide me by link.

thanks in advance.
BR yuji

**Reply by jklj077** [reply]
The "official" explanations on those parameters can be found at the OpenAI API site: <https://platform.openai.com/docs/api-reference/chat/create>. AFAIK, the original document containing their technical definition from OpenAI is lost. If you'd like, you could find the implementation of a specific parameter in open-source inference frameworks, e.g., for `presence_penalty`
- vLLM: <https://github.com/vllm-project/vllm/blob/275de34170654274616082721348b7edd9741d32/vllm/model_executor/layers/utils.py#L95>
- SGLang: https://github.com/sgl-project/sglang/blob/951028968ce1c7e0661a1d06145235787ecef5c3/python/sglang/srt/sampling/penaltylib/presence_penalty.py#L42
